### ISSUE TITLE
Database structure

### ISSUE THREAD

**Issue by gareth-ib** [spatie/laravel-medialibrary#1096]
First off, great library!
But the database structure is too string based for my project with millions and millions of media files.

Would it just take extending the `Spatie\MediaLibrary\Models\Media` in order to get things like `collection_name` and `model_type`, etc... to be reference tables instead of strings in each media row? Or are there dependencies all over? Or any examples of this style I'm talking about?

Also the MediaRepository would need to allow overriding in order to make the data structure be user-specified... potentially mean that the default setup would use a blank `\App\MediaRepository` class that extends the real one. which could stay blank for default usage... or if all the database stuff was moved into a separate optional package, and that could be forked and modified... something along those lines maybe...

Or is the project open to enhancing the shipped structure?

Basically to have all the string fields other than filename be changed into the primary id from a table specific for the modelTypes, collections, mimeTypes, disks, etc...

probably a separate table for file names and types for the responsive json instead of strings like this...

`"{"medialibrary_original":{"urls":["janus___medialibrary_original_350_334.jpg"]}}"`

not to mention the base64 images... far too much giant strings for me to hold in the database

Thanks!

**Comment by gareth-ib** [comment]
quick structure concept...
it's more complex but much more efficient

```
CREATE TABLE media
(
  media_id          serial NOT NULL,
  mediamodeltype_id integer NOT NULL,
  model_id          bigint NOT NULL,
  collection_id     integer NOT NULL,
  name              character varying(255) NOT NULL,
  mediafile_id      integer NOT NULL,
  mime_type_id      integer NOT NULL,
  mediadisk_id      integer NOT NULL,
  manipulations     json NOT NULL,
  custom_properties json NOT NULL,
  responsive_images json NOT NULL,
  order_column      integer,
  created_at        timestamp(0) without time zone,
  updated_at        timestamp(0) without time zone,
  PRIMARY KEY (media_id)
);

CREATE TABLE mime_type (
  mime_type_id  serial NOT NULL,
  name          character varying(255) NOT NULL,
  PRIMARY KEY (mime_type_id),
  UNIQUE (name)
);

CREATE TABLE mediamodeltype (
  mediamodeltype_id serial NOT NULL,
  name              character varying(255) NOT NULL,
  PRIMARY KEY (mediamodeltype_id),
  UNIQUE (name)
);

CREATE TABLE mediacollection (
  mediacollection_id   serial NOT NULL,
  name                  character varying(255) NOT NULL,
  PRIMARY KEY (media_collection_id),
  UNIQUE (name)
);

CREATE TABLE mediafile (
  mediafile_id      serial NOT NULL,
  mediafiletype_id  integer NOT NULL,
  name              character varying(255) NOT NULL,
  size              integer NOT NULL,
  width             integer NOT NULL,
  height            integer NOT NULL,
  PRIMARY KEY (media_file_id),
  UNIQUE (name)
);

-- for things like : medialibrary_original, responsiveImageWithPlaceholder, responsiveImage
CREATE TABLE mediafiletype (
  mediafiletype_id  serial NOT NULL,
  name              character varying(255) NOT NULL,
  PRIMARY KEY (media_file_id),
  UNIQUE (name)
);                
                
CREATE TABLE mediadisk (
  mediadisk_id     serial NOT NULL,
  name              character varying(255) NOT NULL,
  PRIMARY KEY (media_disk_id),
  UNIQUE (name)
);
```

**Comment by gareth-ib** [comment]
but probably ideal if the code was setup so the user can define their own structure

**Comment by freekmurze** [comment]
For the vast majority of users the current table definition is fine. Your proposed structured would make the code more complex with no benefits for most users. That's why I'm going to leave things as they are for now. Simple and pretty usable for the proverbial 80%.

Feel free to fork the code and modify it to your needs.
